### PR TITLE
Add scroll lock when using trackpad on web

### LINKED ISSUE #143276: Web: Scroll axis is not locked during trackpad/touchpad drags (still not fixed?)
### Steps to reproduce

asd

### Expected results

adsf

### Actual results

dsf

### Code sample

<details open><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

### PR DESCRIPTION
**Pull request by IchordeDionysos**
_Related PR: https://github.com/flutter/engine/pull/50645 (Depending on where it should be fixed this or the other PR is obsolete)_

-- 

This is an attempt to fix weird scrolling behavior with trackpads on Flutter web.
It's done by looking at the last scroll events and locking the axis to the major axis of the first scroll axis within a series.
The lock resets automatically if no scroll event is received within x amount of time (e.g. 10ms).

Attempts to fix https://github.com/flutter/flutter/issues/143276

Check https://github.com/flutter/flutter/issues/143276#issuecomment-1950566307 for reasons for introducing this change on web and screen recordings before and after.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/gestures/pointer_signal_resolver.dart [hunk 1]
```diff
@@ -99,14 +131,31 @@ class PointerSignalResolver {
       return;
     }
     assert(_isSameEvent(_currentEvent!, event));
+
+    print(event.kind);
+    // TODO(Ichordedionysos): Re-enable check for trackpad once the event.kind is properly exposing trackpad usage.
+    if (event is PointerScrollEvent /*&& event.kind == PointerDeviceKind.trackpad*/) {
```
Comment (moffatman) [inline packages/flutter/lib/src/gestures/pointer_signal_resolver.dart:137]: Accurate `event.kind` is only possible through my heuristic in certain browsers. But I think it's safe to lock axis on mouse scrolling too unconditionally, usually mouse scrolling is just one axis anyway.

Comment (IchordeDionysos) [reply packages/flutter/lib/src/gestures/pointer_signal_resolver.dart:137]: Yes, I had the same thought 👍

**Comment by moffatman** [comment]
My thinking was, `scrollable.dart` should register a "claim" for this scroll sequence. Then all the scrolls in the current series will be routed to that scrollable, and not go to others. This would let free-axis diagonal scrolling work. While letting individual scrollables lock-out other scrollables. However you do have an issue of who should do the "claim", because you may need a few events to make the judgement clear.

**Review by Piinks — COMMENTED** [review]
Hey @IchordeDionysos what is the status of this PR? It looks like there are some missing tests, and failing existing tests.

Typically we have implemented heuristics for identifying scrolling devices in the engine rather than in the framework. Do you think the linked engine PR is the approach you would like to take?

**Comment by moffatman** [comment]
This can't be fixed in the engine. My heuristic already exists in the engine to identify which scroll events are from trackpad vs mouse wheel. Then it's up to the framework to decide how to handle the events differently. Implementing a scroll axis lock in the engine would break diagonal scrolling as it would destroy off-axis deltas before they reach the framework. Instead I still think the best way to go forward here is to enhance `PointerSignalResolver` with support time-windowed routing locks for discrete scroll events.

**Comment by Piinks** [comment]
Ah thank @moffatman for the context!

**Comment by Piinks** [comment]
(Triage) @IchordeDionysos are you still working on this PR?

**Comment by IchordeDionysos** [comment]
@Piinks I did work on this as a side-project.
I do lack some more fundamental knowledge on how to best achieve only locking scrolls for vertical scrolls.

In the scrollable.dart there seems to be no data if it's a free-axis scrolling or not ...

I wasn't able to spend more time since then!
I can take another look today, if someone of you would like to take it over, that would be fine for me as well.

**Comment by Piinks** [comment]
Yeah, I am not sure the information we need here is available on web right now. I don't know that anyone will be able to pick this PR up right now. You are welcome to continue working on it, otherwise we can close it to return to later. I appreciate your work here, thank you!

**Comment by IchordeDionysos** [comment]
@Piinks were you talking about whether trackpad events are properly exposed on the web or whether the widget is vertically, horizontally, or all-axis scrollable?

I understand that the first is right now not possible due to browser API limitations.

However, the second should be possible (even on the Web), no?
Generally, when I use e.g. a SingleChildScrollView I can specify an axis and this axis information would be interesting when trying to build a heuristic, e.g. to only lock the scroll axis (and only expose scroll events in a single axis) when the scroll container expects a single axis scroll.

**Comment by Piinks** [comment]
> Generally, when I use e.g. a SingleChildScrollView I can specify an axis and this axis information would be interesting when trying to build a heuristic, e.g. to only lock the scroll axis (and only expose scroll events in a single axis) when the scroll container expects a single axis scroll.

The only all-axis scrollable (can scroll in two directions) is TwoDimensionalScrollable, which is used by TwoDimensionalScrollView. SingleChildScrollView can only scroll in one axis, so it does not need to lock one axis or the other, because there is only one.

> were you talking about whether trackpad events are properly exposed on the web

Yup! this is what I was referring to.

**Comment by Piinks** [comment]
Oh and I forgot to mention, the TwoDimensionalScrollable has axis locking options through DiagonalDragBehavior. :)

**Comment by moffatman** [comment]
The issue here is, we need to lock it to the same widget destination. I have an idea how to do it, it's very possible to improve this (not make it perfect but a lot better), I just dont have much effort now to write it, especially not writing tests. you can contact me on flutter discord or something @IchordeDionysos I can help you out with it.

**Pull request closed without merging**